Make Generated Structs Into<Default Struct>

wd40bug opened the pull request:
Just thought that since all the structs contain the same data they should be into the same base struct. This was very useful for my own application.

killzoner commented:
Hello,
Can you stash your commits? I'm reviewing this then, but overall this seems ok to me

killzoner commented:
CI is failing, it seems to break some stuff, probably around handling of prefixes.
If you can look into adding some tests around this feature too, this would be great 🙏

wd40bug commented:
I'll get on that this evening. Thanks 😁

wd40bug committed:
CI Tests passing, new tests added, and changes in a complete-ish form

Added Into Base

preserve tokens around impl

No from over into clippy error

preserve tokens when parsing impl and added option for the into impls

Changes to how ident is detected in item_impl

added tests

wd40bug commented:
Howdy, tests are working on my end.
I had to rework it slightly to account for the possibility of the end user not generating the default struct, so I added `gen_into` as an option of the clappen macro.
Thanks

wd40bug commented on src/clappen/attrs.rs:
Added the gen_into option for this parser

wd40bug commented on src/clappen/mod.rs:
```diff
@@ -129,18 +155,36 @@ pub(crate) fn create_template(
+            #default
+        };
+    }
+
```
It doesn't make sense to keep the same structure if gen_into is used, since creation of the default struct is implied or required (I chose to make it implied since that seemd better)

killzoner replied:
We might add a different default behaviour behind a feature if that's more convenient to you?
Like if the feature is activated this is generated by default? Probably later 😄

wd40bug commented on src/clappen_impl/item_impl.rs:
```diff
@@ -22,12 +23,56 @@ impl ProcessItem for ItemImpl {
             attrs_prefix.as_str(),
         ]));
 
+        // Preserve surrounding impl tokens
```
The code originally discarded tokens around the impl which led to issues when implementing traits. Additionally, identifiers were treated without respect to the possibility of containing generics

wd40bug commented on tests/expandtest.rs:
```diff
@@ -1,7 +1,7 @@
 fn expandtest() {
     let t = trybuild::TestCases::new();
-    t.pass("tests/**/*.pass.rs");
+    t.pass("tests/expand/**/*.pass.rs");
```
Otherwise it was running my tests out of place

killzoner replied:
Mmh I'm gonna look into this, i'd like to not create folder specific handling

killzoner replied:
Fixed

killzoner commented on tests/into/fail.rs:
```diff
@@ -0,0 +1,8 @@
+}
+// This macro doesn't work with no argument when gen_into is provided
+fn main() {
+    generate!()
```
I think you're missing a `;` here for the actual expected error

killzoner commented on tests/into.rs:
```diff
@@ -0,0 +1,39 @@
+#[test]
+fn into_fail() {
+    let t = trybuild::TestCases::new();
+    t.compile_fail("tests/into/*.rs");
```
Can you add an expand test?

killzoner commented on tests/into.rs:
```diff
@@ -0,0 +1,39 @@
+    let one_into: SimpleStruct = one.into();
+    let two_into: SimpleStruct = two.into();
+    assert_eq!(one_into, compare);
+    assert_eq!(two_into, compare);
```
Did you test with nested structs? Ideally i'd like to have a test case with nested structs using both default prefix and field prefix. I can do it later after merge though

killzoner commented:
Nice work ! I've commented some bits, and probably this would need a rebase for the CI to run.
If everything is ok, I'll merge and release early next week as I'm off without my laptop right now

killzoner commented:
Hey i'm gonna merge this to a dedicated branch and do the modifications, tagging you on the final merge if that's ok for you

killzoner committed:
Merge branch 'base-into' into into-base

killzoner commented:
I don't think it's gonna work with the current implementation because as long as you go to nested fields, you need to `into()` also for them.
I'm gonna look into this when I have some time

killzoner commented:
See https://github.com/killzoner/clappen/commit/15ef5623983cfcd16dfef0322be12b8c74be7cd6#diff-5d5ce0de3e74ea48629bb2333951485453363f8a55ef7ff24dde648fc3477c7bR43 where default struct is holding field prefix and default prefix, we would need to rethink some part of this (maybe by adding a trait bound to the base struct ?), but this won't work as is

wd40bug commented:
Yeah I ran into that yesterday. I was gonna modify the macros to keep track of all the prefixes separately (passed in an array), which should fix the issue. The heart of the issue is that the prefixed struct fields in a prefixed structure (i.e. the nested fields) aren't Into the prefixed field in the first struct

wd40bug commented:
Should keep the user facing API the same, but will slightly change the internal api
